lib/web/ui/join_token.go

```diff
@@ -0,0 +1,73 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package ui
+
+import (
+	"strings"
+	"time"
+
+	"github.com/gravitational/teleport/api/types"
+)
+
+type JoinToken struct {
+	// Id is the name of the token
+	Id string `json:"id"`
+	// SafeName returns the name of the token, sanitized appropriately for
+	// join methods where the name is secret.
+	SafeName string `json:"safeName"`
+	// Expiry is the time that the token resource expires
+	Expiry time.Time `json:"expiry"`
+	// Roles are the roles granted to the token
+	Roles types.SystemRoles `json:"roles"`
+	// IsStatic is true if the token is statically configured
+	IsStatic bool `json:"isStatic"`
+	// Method is the join method that the token supports
+	Method types.JoinMethod `json:"method"`
+	// SuggestedLabels contains the set of labels we expect the node to set when using this token
+	SuggestedLabels []Label `json:"suggestedLabels,omitempty"`
+	// AllowRules is a list of allow rules
+	AllowRules []string `json:"allowRules,omitempty"`
+}
+
+func MakeJoinToken(token types.ProvisionToken) JoinToken {
+	labels := token.GetSuggestedLabels()
+	suggestedLabels := make([]Label, len(labels))
+
+	for labelKey, labelValues := range labels {
+		suggestedLabels = append(suggestedLabels, Label{
+			Name:  labelKey,
+			Value: strings.Join(labelValues, " "),
+		})
+	}
+
+	return JoinToken{
+		Id:              token.GetName(),
+		SafeName:        token.GetSafeName(),
+		Expiry:          token.Expiry(),
+		Roles:           token.GetRoles(),
+		IsStatic:        token.Expiry().Equal(time.Unix(0, 0).UTC()),
```

Comment:
I'm not sure about this one.
A static token is a token that is configured directly into the `teleport.yaml` in the auth service. Those exist inside the type `types.StaticTokensV2`

Which is different from a token with no expiration, which you can create dynamically.
So while static tokens don't have expiration, not all `Expiry == 0` are static tokens.

Reply:
Static tokens are set from the config with zero Unix time (1970-01-01), but if I manually configure a token with no ttl, it's actually set to zero time (0001-01-01) so it's not an ideal distinction but it does work. 

Ideally I wanted to use the Origin but it seems the origin for all tokens (in my testing) just return empty string.

Reply:
TIL, thank you.
Can we convert this check into a token's method? I think this distinction should be really close to the token's definition.

Reply:
Do you mean something like making JoinToken.IsStaic() ?

Reply:
Yes